pesudocode/voxel_grid_structure_3.hpp

```diff
@@ -0,0 +1,137 @@
+#include <iostream>
+#include <type_traits>
+#include <vector>
+#include <unordered_map>
+#include <Eigen/Dense>
+#include <iterator>
+
+class PointXYZ
+{
+public:
+    PointXYZ() {}
+};
+
+template <typename PointT>
+class PointCloudT
+{
+public:
+    PointCloudT() {}
+};
+
+template <typename PointT>
+class Filter
+{
+    using PointCloud = PointCloudT<PointXYZ>;
+
+public:
+    Filter() {}
+    virtual void applyFilter(PointCloud &pc) = 0;
+    void filter(PointCloud &output) { applyFilter(output); }
+};
+
+// ================== Base of VoxelGrid ==================
+
+// Check the point type:
+
+// Option 1
+// Read the point type based on the returning type of reduceGrid(...)
+// With this implementation, I think we can't check beforehand if
+// the user provided the mandatory functions
+// (use invoke_result_t from C++17)
+#define GET_POINT_TYPE_1(GridStruct, Grid) \
+    typename std::result_of_t<decltype (&GridStruct::reduceGrid)(GridStruct, const Grid &)>
+
+// Option 2
+// Check typedef/using in the GridStruct
+// Require user to define it
+#define GET_POINT_TYPE_2(GridStruct) \
+    typename GridStruct::PointType
+
+template <typename GridStruct, typename Grid, typename PointT = GET_POINT_TYPE_2(GridStruct)>
```

Comment:
Why is  `Grid` a template parameter if it's used in just `std::unordered_map<stze_t, Grid>`? As such, can't we just encapsulate the entire unordered_map in a template parameter?

In L76, we can just pass in grid instead of grid.second. Let the user deal with it

Reply:
I have tried that also, but I feel like in such case we might enforce too many rules to the user

Reply:
But after thinking more on it, I agree with you that we should provide more flexibility to the users